pkg/apis/kubeone/validation/validation.go

```diff
@@ -352,17 +352,45 @@ func ValidateContainerRuntimeConfig(cr kubeoneapi.ContainerRuntimeConfig, versio
 func ValidateClusterNetworkConfig(c kubeoneapi.ClusterNetworkConfig, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	if len(c.PodSubnet) > 0 {
-		if _, _, err := net.ParseCIDR(c.PodSubnet); err != nil {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("podSubnet"), c.PodSubnet, ".clusterNetwork.podSubnet must be a valid CIDR string"))
-		}
+	subnetCountErr := func(node, subnet string, count int, ipFamily kubeoneapi.IPFamily) *field.Error {
+		return field.Invalid(fldPath.Child(node), subnet, fmt.Sprintf(".clusterNetwork.%s must specify %d subnet(s) for %q IP family.", node, count, ipFamily))
+	}
+
+	invalidFamilyErr := func(node, subnet string, ipFamily kubeoneapi.IPFamily) *field.Error {
+		return field.Invalid(fldPath.Child(node), subnet, fmt.Sprintf(".clusterNetwork.%s must be valid %q subnet.", node, ipFamily))
 	}
-	if len(c.ServiceSubnet) > 0 {
-		if _, _, err := net.ParseCIDR(c.ServiceSubnet); err != nil {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("serviceSubnet"), c.ServiceSubnet, ".clusterNetwork.serviceSubnet must be a valid CIDR string"))
+
+	var subnetValidators []func(string) bool
+	switch c.IPFamily {
+	case kubeoneapi.Unspecified, kubeoneapi.IPv4:
+		subnetValidators = append(subnetValidators, netutils.IsIPv4CIDRString)
+	case kubeoneapi.IPv6:
+		subnetValidators = append(subnetValidators, netutils.IsIPv6CIDRString)
+	case kubeoneapi.DualStack:
+		subnetValidators = append(subnetValidators, netutils.IsIPv4CIDRString, netutils.IsIPv6CIDRString)
+	default:
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("ipFamily"), c.IPFamily, "unknown ipFamily"))
+	}
+
+	validateCIDRs := func(node, subnet string) {
+		if len(subnet) == 0 {
+			return
+		}
+		subnets := strings.Split(subnet, ",")
+		if len(subnets) != len(subnetValidators) {
+			allErrs = append(allErrs, subnetCountErr(node, subnet, len(subnetValidators), c.IPFamily))
+		} else {
+			for i, isValid := range subnetValidators {
+				if !isValid(subnets[i]) {
+					allErrs = append(allErrs, invalidFamilyErr(node, subnet, c.IPFamily))
+				}
+			}
```

Comment:
This is implying that you must first provide IPv4 then IPv6 CIDR (when you use dual-stack). This behavior is controlled by the order of validators. That being said, if someone would for any reason change the order of validators, that would break the API.

Does it make sense to introduce two new fields for IPv6 PodSubnet and ServiceSubnet instead of reusing existing fields?

Reply:
> This is implying that you must first provide IPv4 then IPv6 CIDR (when you use dual-stack). This behavior is controlled by the order of validators. That being said, if someone would for any reason change the order of validators, that would break the API.

Right. That is what is expected. 

The other way round isn't supported. It might never be supported since dualstack is just a step on the way to going full IPv6. 

> Does it make sense to introduce two new fields for IPv6 PodSubnet and ServiceSubnet instead of reusing existing fields?

We could do it that way. KKP, MC, KubeOne does it like this tho.

Reply:
I think in kubeone case we could actually support both `IPv4+IPv6` (IPv4 as primary) and `IPv6+IPv4` case (IPv6 as primary). The main reason why we don't support the IPv6+IPv4 case in KKP yet is that control plane needs to run on IPv6, which is more tricky in KKP's remote control plane case. But should be fairly straightforward in kubeone case.

Reply:
With regard to what @rastislavs said, having two dedicated fields for IPv4 and IPv6 subnets might make sense. I think it would be more intuitive to choose between those two options (`IPv4+IPv6` or `IPv6+IPv4`) via some API field explicitly (e.g. via `IPFamily`), rather than by properly ordering subnets.